### DISCUSSION TITLE
Possible to use a generic 34-key layout on a Glove80?

### DISCUSSION THREAD

**Discussion by rieje** [urob/zmk-config#72] [Q&A] [answered]
Is it possible to use a generic 34-key layout on a Glove80? I only care/develop for a 34-key layout using it on a Glove80 (15 keys and the two bottom thumb keys closest to the thumb on each side). I would like to maintain a single .keymap file and have it usable for multiple split boards. Is this what [ZMK_APPLY_MATRIX_TRANFORM](https://github.com/urob/zmk-helpers/blob/main/docs/core_helpers.md#zmk_apply_matrix_transform) does?

If it's trivial to do, an example would be much appreciated. Otherwise I guess setting `&none` to all the extra keys defined in glove80.keymap is the next best approach.

Thanks.

**Comment by urob** [comment, marked as answer]
Yes, in principle that's possible. You'd need to first define your own matrix transform. Once defined, you can then apply it in your keymap using `ZMK_APPLY_MATRIX_TRANSFORM`.

Out of curiosity, what about the modular approach in this repo doesn't work for you? Effectively, I'm only maintaining a single 34-keys keymap file (`base.keymap`) which is applied to all my keyboards of different size, including my Glove80.

**Reply by rieje** [reply]
For now I just want to get rid of the extra `X_*` keys in the keymap because I need to reduce font size by a lot to view the keymap without wrapping lol. 

EDIT:

I copied the original matrix for glove80 then made [this edit](https://cdn.discordapp.com/attachments/1269860291147989084/1269871111647268926/image.png?ex=66b1a322&is=66b051a2&hm=0c77ce1210788c6d35e5c07234f30167a9fa5298e128141ed30dc669827cd5a6&) along with changing rows to `<4>`. I put this in [glove80.keymap](https://github.com/urob/zmk-config/blob/main/config/glove80.keymap):

    / {
      chosen {
        zmk,kscan = &kscan0;
        zmk,matrix-transform = &default_transform;
        zephyr,code-partition = &code_partition;
        zephyr,sram = &sram0;
        zephyr,flash = &flash0;
      };

      glove80_transform: keymap_transform_glove80 {
        compatible = "zmk,matrix-transform";
        columns = <14>;
        rows = <4>;
        map = <
          RC(2,1) RC(2,2) RC(2,3) RC(2,4) RC(2,5)                                                    RC(2,8) RC(2,9) RC(2,10) RC(2,11) RC(2,12)
          RC(3,1) RC(3,2) RC(3,3) RC(3,4) RC(3,5)                                                    RC(3,8) RC(3,9) RC(3,10) RC(3,11) RC(3,12)
          RC(4,1) RC(4,2) RC(4,3) RC(4,4) RC(4,5)                                                    RC(4,8) RC(4,9) RC(4,10) RC(4,11) RC(4,12)
          RC(5,1) RC(5,2) RC(5,3) RC(5,4)         RC(3,6) RC(4,6)                   RC(4,7)  RC(3,7)         RC(5,9) RC(5,10) RC(5,11) RC(5,12)
        >;
      };
    };

then removed all the `define X_` lines. It builds (it fails to build if I add `#include "zmk-helpers/helper.h"` to the beginning of this file and then `ZMK_APPLY_MATRIX_TRANSFORM(keymap_transform_glove80)` after the above code where it reports error `parse error: expected label reference (&foo)` at [this line in base.keymap](https://github.com/urob/zmk-config/blob/main/config/base.keymap#L184).

After flashing it, the F keys of the Glove80 is where  the top of my 34-key base layer is. It seems the base layer shifted up 2 rows and as least some other unexpected changes. Combos seem to be in the correct places though.

Do you know if I need to edit something else in the config [like here](https://github.com/urob/zmk-config/blob/main/config/base.keymap#L51)? And can I drop any includes once this works?

Much appreciated.

**Reply by urob** [reply]
A bit hard to say much without seeing your code.  Where did you add the transform? For the transform to work, you'll need additional headers. Checking out the original firmware files for the glove (or any other board) should give you guidance here. (Also make sure you don't have name conflicts between the various macros defined by those headers and your keymap, e.g., `RC`)

I'm also not sure I understand what exactly went wrong. How did you manage to flash your keyboard if you got a compile error when building your keymap? In either case, if the F-keys are somewhere on flashed board, it sounds like it's not reading your keymap correctly (assuming you don't have F-keys on the base layer of your 34-key map).

Finally, you want to make sure to build with the `-p` (pristine) flag, so the build cache doesn't interfere with your matrix transform.

**Reply by rieje** [reply]
Appreciate the response. I added the `-p` flag and had done some changes. [base.keymap](https://0x0.st/XVTM.txt), [glove80.keymap](https://0x0.st/XVTu.txt).

Changes:

For `glove80.keymap` it's basically a copy of [this](https://github.com/zmkfirmware/zmk/blob/main/app/boards/arm/glove80/glove80.dtsi) with the matrix modified with using [this suggestion](https://media.discordapp.net/attachments/1269860291147989084/1269871111647268926/image.png?ex=66b24be2&is=66b0fa62&hm=1308229b1f9f3b28a871c00799770385ae67fdf98695b0c09f489eb6b453451b&=&format=webp&quality=lossless&width=1108&height=258) I also changed the values of the column and row to 14 and 4 respectively. I'm guessing this file is the culprit. I replaced `#include "zmk-helpers/key-labels/glove80.h"` with `#include "zmk-helpers/key-labels/34.h"`. I first tried `ZMK_APPLY_MATRIX_TRANSFORM` with the necessary headers but got some compile error that's probably not related to using the macro.


Problems:

* Left side of 34-key base layout works (rest of the non-34 key layout keys are removed by the matrix as intended), but for some reason holding leftmost left thumb button accesses the `fn` layer, not the `nav` layer and the holding rightmost left thumb button accesses the `nav` layer, not the `sym` layer. There are likely more problems but they were not present before I made changes for the matrix. 

* Right side of board somehow seems is like a copy of the left side of the board with some keys being offset including the bottom *three* right thumb keys being `,` `i` and `k` (for a 34-key layout I only define the 2 closest bottom thumb buttons). So the matrix is definitely wrong.

**Reply by urob** [reply]
- `ZMK_APPLY_MATRIX_TRANSFORM` sets `zmk,matrix-transform = &glove80_transform;` which you already do. So no need for that
- Yes, `34.h` is the right header. No need to re-import the helpers header as it is already imported in base
- No need to redefine all the nodes from the default glove firmware. Cleaning up a bit, `glove80.keymap` could be as simple as:
```c
#include "zmk-helpers/key-labels/34.h"
#include <dt-bindings/zmk/matrix_transform.h>

/* 34-key layout on the Glove80, adapted from original matrix */
// See: https://github.com/zmkfirmware/zmk/blob/main/app/boards/arm/glove80/glove80.dtsi#L20-L32 */
/ {
  &chosen {
    zmk,matrix-transform = &glove80_transform;
  };

  glove80_transform: keymap_transform_glove80 {
    compatible = "zmk,matrix-transform";
    columns = <12>;
    rows = <4>;
    map = <
      RC(2,1) RC(2,2) RC(2,3) RC(2,4) RC(2,5)  RC(2,8) RC(2,9) RC(2,10) RC(2,11) RC(2,12)
      RC(3,1) RC(3,2) RC(3,3) RC(3,4) RC(3,5)  RC(3,8) RC(3,9) RC(3,10) RC(3,11) RC(3,12)
      RC(4,1) RC(4,2) RC(4,3) RC(4,4) RC(4,5)  RC(4,8) RC(4,9) RC(4,10) RC(4,11) RC(4,12)
                              RC(3,6) RC(4,6)  RC(4,7) RC(3,7)
    >;
  };
};

/* activate bluetooth */
#define CONFIG_WIRELESS 1

/* source the main keymap */
#include "base.keymap"
```

- I'm not sure why the two thumb buttons are switched. Are you sure you re-build correctly? If you keep getting the issue, you could try switching the two thumbs in the transform above (though you really shouldn't need to)

- Re: right half, did you adjust the `col-offset` for your transform? You'll want to set it to `7` just as is done for the `default_transform` here:
https://github.com/zmkfirmware/zmk/blob/cd25c12ce99d1ee708d76854a79980c7fb48a331/app/boards/arm/glove80/glove80_rh.dts#L82-L84
 (Make sure this is only triggered when building the right half, otherwise your left half gets shifted as well). If you don't want to mess with that, you might also be able to just overwrite the `default_transform` like so and don't need to worry about setting `chosen` and the `col-offset`:
```c
/ {
  &default_transform {
    map = <
      RC(2,1) RC(2,2) RC(2,3) RC(2,4) RC(2,5)  RC(2,8) RC(2,9) RC(2,10) RC(2,11) RC(2,12)
      RC(3,1) RC(3,2) RC(3,3) RC(3,4) RC(3,5)  RC(3,8) RC(3,9) RC(3,10) RC(3,11) RC(3,12)
      RC(4,1) RC(4,2) RC(4,3) RC(4,4) RC(4,5)  RC(4,8) RC(4,9) RC(4,10) RC(4,11) RC(4,12)
                              RC(3,6) RC(4,6)  RC(4,7) RC(3,7)
    >;
  };
};
```

**Reply by rieje** [reply]
@Taking the suggestions with the following [glove80.keymap](https://0x0.st/XVQw.txt), no offset issues anymore and everything seems fine involving tapping on the base layer. AFAIK this is the only problem now:

* Bottom index finger keys do not trigger FN layer on hold
* Leftmost left thumb key triggers FN layer instead of NAV layer on hold
* Rightmost left thumb key triggers NAV layer instead of SYM layer on hold
* Leftmost right thumb key triggers SYM layer instead of NUM layer on hold

I'm not sure why the issues are with holding these keys to access layers. There's no issues with holding the smart-shift to behave as regular shift and combos also work as expected. Swapping the thumb keys just swaps from leftmost left thumb key's Space-FN to Enter-NAV, i.e. for some reason Space-FN are hold-taps and Enter-NAV are hold-tabs despite the base.keymap defining them as Space-NAV, Enter-SYM (regardless of swapping the thumbs to `RC(4,6) RC(3,6)`). Also haven't found the key that triggers the NUM layer. To be clear, everything that's a tap on base layer including the thumb keys are mapped correctly but the issues above all involve holding for a layer aren't mapped correctly for some reason. The [base.keymap](https://0x0.st/XVTM.txt) remains unchanged and builds were built with `-p` prestine option.

**Reply by urob** [reply]
Re: layer issues. You want to double check your keymap file. I can see at least two issues:
- You are reusing the `sym` label for multiple layers
- Your layers are out of sync with the macro definitions at the beginning of the file. E.g., `num` is the first layer (counting from 0), whereas `NAV=2`
